### PR TITLE
ARROW-6314: [C++] Implement IPC message format alignment changes, provide backwards compatibility and "legacy" option to emit old message format

### PR DESCRIPTION
**Pull request by wesm**
This also moves the alignment multiple to `IpcOptions` and adds the `IpcOptions` argument to more functions.

### CODE REVIEW

File: cpp/src/arrow/ipc/message.cc [hunk 1]
```diff
@@ -227,7 +231,7 @@ std::string FormatMessageType(Message::Type type) {
 Status ReadMessage(int64_t offset, int32_t metadata_length, io::RandomAccessFile* file,
                    std::unique_ptr<Message>* message) {
   ARROW_CHECK_GT(static_cast<size_t>(metadata_length), sizeof(int32_t))
-      << "metadata_length should be at least 4";
+      << "metadata_length should be at least 8";
```
Comment (pitrou) [inline cpp/src/arrow/ipc/message.cc]: Hmm... the comperand hasn't changed, so why the change in the error message?

File: cpp/src/arrow/ipc/message.cc [hunk 2]
```diff
@@ -237,15 +241,31 @@ Status ReadMessage(int64_t offset, int32_t metadata_length, io::RandomAccessFile
                            " metadata bytes but got ", buffer->size());
   }
 
-  int32_t flatbuffer_size = *reinterpret_cast<const int32_t*>(buffer->data());
+  const int32_t continuation = util::SafeLoadAs<int32_t>(buffer->data());
+  int32_t message_length = 0;
+  int32_t prefix_size = -1;
+  if (continuation == kIpcContinuationToken) {
+    // Valid IPC message, parse the message length now
+    message_length = util::SafeLoadAs<int32_t>(buffer->data() + 4);
```
Comment (pitrou) [inline cpp/src/arrow/ipc/message.cc]: It's not obvious whether the buffer can be too small here.

File: cpp/src/arrow/ipc/message.cc [hunk 3]
```diff
@@ -237,15 +241,31 @@ Status ReadMessage(int64_t offset, int32_t metadata_length, io::RandomAccessFile
...
-  if (flatbuffer_size + static_cast<int>(sizeof(int32_t)) > metadata_length) {
-    return Status::Invalid("flatbuffer size ", metadata_length,
+  if (message_length + prefix_size != metadata_length) {
+    return Status::Invalid("flatbuffer size ", message_length,
                            " invalid. File offset: ", offset,
                            ", metadata length: ", metadata_length);
   }
 
-  auto metadata = SliceBuffer(buffer, 4, buffer->size() - 4);
+  auto metadata = SliceBuffer(buffer, prefix_size, buffer->size() - prefix_size);
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/ipc/message.cc]: We would be copying only if `prefix_size == 4`, right?

File: cpp/src/arrow/ipc/message.cc [hunk 4]
```diff
@@ -237,15 +241,31 @@ Status ReadMessage(int64_t offset, int32_t metadata_length, io::RandomAccessFile
                            " metadata bytes but got ", buffer->size());
   }
 
-  int32_t flatbuffer_size = *reinterpret_cast<const int32_t*>(buffer->data());
+  const int32_t continuation = util::SafeLoadAs<int32_t>(buffer->data());
+  int32_t message_length = 0;
```
Comment (pitrou) [inline cpp/src/arrow/ipc/message.cc]: I get lost. Is this the message body length? The actual metadata length?

**Review by pitrou — COMMENTED** [review]
LGTM overall.

**Comment by pitrou** [comment]
I fixed the CUDA issue.

**Comment by kszucs** [comment]
@ursabot build
